src/Build/Graph/GraphBuilder.cs

```diff
@@ -621,9 +621,42 @@ internal class GraphEdges
                     ErrorUtilities.VerifyThrow(ReferenceItems.TryGetValue(key, out ProjectItemInstance referenceItem), "All requested keys should exist");
                     return referenceItem;
                 }
+            }
+
+            public void AddOrUpdateEdge((ProjectGraphNode node, ProjectGraphNode reference) key, ProjectItemInstance edge)
+            {
+                ReferenceItems.AddOrUpdate(
+                    key,
+                    addValueFactory: static ((ProjectGraphNode node, ProjectGraphNode reference) key, ProjectItemInstance referenceItem) => referenceItem,
+                    updateValueFactory: static ((ProjectGraphNode node, ProjectGraphNode reference) key, ProjectItemInstance existingItem, ProjectItemInstance newItem) =>
+                    {
+                        string existingTargetsMetadata = existingItem.GetMetadataValue(ItemMetadataNames.ProjectReferenceTargetsMetadataName);
+                        string newTargetsMetadata = newItem.GetMetadataValue(ItemMetadataNames.ProjectReferenceTargetsMetadataName);
+
+                        // Bail out of the targets are the same.
+                        if (existingTargetsMetadata.Equals(newTargetsMetadata, StringComparison.OrdinalIgnoreCase))
+                        {
+                            return existingItem;
+                        }
 
-                // First edge wins, in accordance with vanilla msbuild behaviour when multiple msbuild tasks call into the same logical project
-                set => ReferenceItems.TryAdd(key, value);
+                        existingTargetsMetadata = GetEffectiveTargets(existingItem, existingTargetsMetadata);
+                        newTargetsMetadata = GetEffectiveTargets(newItem, newTargetsMetadata);
+
+                        ProjectItemInstance mergedItem = existingItem.DeepClone();
```

Comment:
In some (but not all :( ) cases, this is one of the objects returned from a `ProjectInstance.GetItems(...)` call, so I'm wanting to avoid mutating the value.

It's kinda weird to hold onto the `ProjectItemInstance` in the first place when we need very little from it, but I suppose it's a perf optimization to avoid allocating yet another object to hold a mutable subset of the `ProjectItemInstance`.